Implement irc commands

JonathanDUFOUR opened the pull request:
It is related to #12.
All the listed commands are not implemented yet, but the following are so far:
- PASS
- NICK
- USER
- OPER
- AWAY
- MODE

However, the unit tests for the above commands are not implemented yet.

JonathanDUFOUR committed:
Replace `StatusCode` by exceptions                      [OK]
Implement the IRC commands                              [DOING]
  MODE                                                  [OK]
  QUIT                                                  [TODO]
  JOIN                                                  [TODO]
  PART                                                  [TODO]
Test the IRC commands                                   [TODO]
Update the RFC summary                                  [DOING]
  List the command steps with <ol> tags                 [DOING]
    QUIT                                                [TODO]
    JOIN                                                [TODO]
    PART                                                [TODO]
  List the possible assertion failures and how to react [DOING]
    JOIN                                                [TODO]
    PART                                                [TODO]

JonathanDUFOUR committed:
Merge branch 'main' into implement_irc_commands

JonathanDUFOUR committed:
Clean the code                                           [OK]
  Remove unused includes                                 [OK]
  Remove unused methods/functions                        [OK]
  Add @throw sections for C++ standard library functions [OK]

JonathanDUFOUR committed:
Changed the away mode to a field, not a flag [OK]

JonathanDUFOUR committed:
Merge branch 'main' into implement_irc_commands

JonathanDUFOUR committed:
Correct the RFC summary about the format of the replies  [OK]
Switch `format_reply()` into `Client::formatted_reply()` [OK]
Do not disconnect clients directly in the IRC commands   [OK]
Mark them with a dedicated flag instead                  [OK]
Fix `Nickname::is_valid()`                               [OK]
Handle SIGINT to stop the server properly                [OK]
Add a timeout to `epoll()`                               [OK]
Refactor the server routine                              [OK]
Close epoll socket in `Server::~Server()`                [OK]

JonathanDUFOUR committed:
Merge branch 'main' into implement_irc_commands

JonathanDUFOUR commented:
Since I opened this merge request, I have refactored many things around in the project.
If you could review all those changes with high attention please, I may have forgotten some details not to merge into the `main` branch.

Any report/remark is welcome.

Saphirac commented on Makefile:
```diff
@@ -3,62 +3,132 @@
- LINK = ${shell which clang++}
-MKDIR =	${shell which mkdir} -p
-   RM = ${shell which rm} -rf
+   AR := ${shell which ar} rcs
```
What's the use of these : ?

JonathanDUFOUR replied:
I switched from operator `=` to operator `:=` because after talking about it with a friend from 42, I realized that the operator `:=` is more efficient than the operator `=`. Indeed, the operator `=` use more memory and instructions under the hood, because when you assign a variable using the operator `=`, for example:
```make
ORIGIN = foo
  COPY = ${ORIGIN}
```
then `make` will keep track that the variable `COPY` depends on the variable `ORIGIN`.
It means that if you do the following:
```make
ORIGIN = foo
  COPY = ${ORIGIN}
...
ORIGIN = bar
```
then the variable `COPY` will have the value `bar` in.
On the other hand, using the operator `:=` allows you to assign a variable without keeping track of the possible dependencies.
It means that if you do the following:
```make
ORIGIN = foo
  COPY := ${ORIGIN}
...
ORIGIN = bar
```
then the variable `COPY` will have the value that it had at the moment of its last assignation, which is `foo`.

Therefore, using the operator `:=` as much as possible is a better practice because it costs less memory and instructions under the hood.

Saphirac replied:
Thank you for the explanation, it does seem more efficient to use := for the Make

Saphirac commented on Makefile:
```diff
@@ -3,62 +3,132 @@
 				} \
-				${addprefix Channel/, \
-					Channel \
+				${addprefix exception/, \
```
I really said that I didn't want to have multiple files for every exceptions...why do you not accept my choices on that ?

JonathanDUFOUR replied:
Because we recently talked about how we shall organize the files related to our classes, and we both agreed to make, for every class, a folder named the same as the class itself, and to put the constructors + destructor + accessors + mutators of the class in a first file named `core.cpp` inside that folder, and to put the rest of the methods of the class in a second file named `methods.cpp` inside this folder too.
And our exceptions are classes.

Saphirac commented on Makefile:
```diff
@@ -94,32 +164,53 @@ LIB_SRC = \
-					UserModeMask \
-					Username \
+				${addprefix specialized_string/, \
+					${addprefix ChannelName/, \
```
Maybe we don't need a core file + a method file for Class that have only ONE method...

JonathanDUFOUR replied:
I don't want multiple standards and exceptions to our defined rules.

Saphirac commented on Makefile:
```diff
@@ -130,32 +221,31 @@ CXXFLAGS = \
 
 ifeq (${DEBUG}, 1)
-	CXXFLAGS += -g -DDEBUG=1
+	CXXFLAGS += -gdwarf-4 -DDEBUG=1
```
Use?

JonathanDUFOUR replied:
This is because I encountered problems related to the `dwarf` version when compiling with `-g` (`clang` compiles with a version that valgrind doesn't support yet by default). Doing this that way doesn't change anything at use.

Saphirac commented on private/ChannelMode.hpp:
```diff
@@ -3,24 +3,26 @@
-	Key,
-	Limit,
-	NoMessagesFromOutside,
-	ChannelOperator,
```
Why adding it since the creator is necessarily operator ?

JonathanDUFOUR replied:
The `ChannelOperator` is needed to add and remove channel operators using the `Channel::set_mode(ChannelOperator, nickname)` and the `Channel::clear_mode(ChannelOperator, nickname)` methods. Also, it is required to know whether a channel has channel operators using the `Channel::is_set(ChannelOperator)` method. Indeed, it is possible not to have any channel operator in a channel.

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+		this->_bits &= ~(1 << IsAboutToBeDisconnected);
+		return;
+	default:
+		throw NotAFlag();
```
Isn't a whole exception for that a bit too much ?

JonathanDUFOUR replied:
Same answer as here: https://github.com/Saphirac/ft_irc/pull/16/files/9aa8dfb6f043550622db30ffb3d12aa3f400efea#r1516700940

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+		this->_bits |= 1 << IsAboutToBeDisconnected;
+		return;
+	default:
+		throw NotAFlag();
```
Isn't a whole exception for that a bit too much ?

JonathanDUFOUR replied:
I don't think so. This is an error caused by the wrong usage of the method. In such a case, we want to alert the developer as soon as possible of their mistake, and we won't ever want to let it compile and run with such a wrong usage. Therefore, I think that throwing an exception in that case is relevant.

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+	case IsAboutToBeDisconnected:
+		return (this->_bits & 1 << IsAboutToBeDisconnected) != 0;
+	default:
+		throw NotAFlag();
```
Isn't a whole exception for that a bit too much ?

JonathanDUFOUR replied:
Same answer as here: https://github.com/Saphirac/ft_irc/pull/16/files/9aa8dfb6f043550622db30ffb3d12aa3f400efea#r1516700940

JonathanDUFOUR committed:
Fix the `Key::is_valid()` method [OK]
Update the RFC summary                                  [DOING]
  List the command steps with <ol> tags                 [DOING]
    QUIT                                                [OK]
    JOIN                                                [OK]
    PART                                                [TODO]
  List the possible assertion failures and how to react [DOING]
    JOIN                                                [OK]
    PART                                                [TODO]

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+void Client::append_to_msg_in(std::string const &s) { this->_msg_in += s; }
+
+#define TERMINATING_SEQUENCE        "\r\n"
+#define TERMINATING_SEQUENCE_LENGTH sizeof(TERMINATING_SEQUENCE) - 1
```
why the -1?

JonathanDUFOUR replied:
Because the `sizeof` keyword computes the actual size of the given literal string as a memory area. Therefore, because the literal string `"\r\n"` is represented as the array `['\r', '\n', '\0']`, which is composed of 3 bytes, `sizeof("\r\n")` will be 3.

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+ *
+ * @throw `std::exception` if a function of the C++ standard library critically fails.
+ */
+void Client::append_to_msg_out(std::string const &msg) { this->_msg_out += msg + "\r\n"; }
```
You made a TERMINATINGSEQUENCE define why not using it here ?

JonathanDUFOUR replied:
Good notice, I will change that.

Saphirac commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+ * @throw `InvalidConversion` if a conversion specification is invalid.
+ * @throw `std::exception` if a function of the C++ standard library critically fails.
+ */
+std::string Client::formatted_reply(int const reply_number...) const
```
what was wrong with the old format reply?

Saphirac commented on src/lib/class/Server/command/cap.cpp:
```diff
@@ -3,35 +3,32 @@
 {
 	if (parameters.empty())
-		sender.append_to_msg_out(format_reply(ERR_NEEDMOREPARAMS, "CAP"));
+		return sender.append_to_msg_out(sender.formatted_reply(ERR_NEEDMOREPARAMS, "CAP"));
```
if format reply is always used like that maybe it can automatically append ?
Is it normal to return it?

JonathanDUFOUR replied:
Making the `Client::formatted_reply` method automatically append the newly generated message to the output buffer of the client is not a bad idea. You should open an issue for that.
Concerning the return, yes, it is wanted to return. By doing so, I can stop the function here, and I can `return Client::append_to_msg_out()` because it also returns a `void`.

Saphirac commented on src/lib/class/Server/command/cap.cpp:
```diff
@@ -3,35 +3,32 @@
-	if (parameters.size() > 1 && parameters[0] == "LS" && parameters[1] == "302")
-		sender.append_to_msg_out(':' + this->_name + "CAP * LS :none\n\r");
-	return Success;
+	if (parameters[0] == "LS")
```
why the 302 delete?

JonathanDUFOUR replied:
I deleted the 302 because it was the version number of the latest capability negotiation protocol supported by the client. And as long as we don't support any capabilities, we don't care about it.

JonathanDUFOUR committed:
Update the RFC summary                             [DOING]
  List the command steps using lists               [DOING]
    PART                                           [OK]
    TOPIC                                          [TODO]
  List the possible assertion failures using lists [DOING]
    PART                                           [OK]
    TOPIC                                          [TODO]

JonathanDUFOUR committed:
Update the RFC summary                 [DOING]
  List the command steps               [DOING]
    TOPIC                              [OK]
    NAMES                              [TODO]
  List the possible assertion failures [DOING]
    TOPIC                              [OK]
    NAMES                              [TODO]

Saphirac commented on src/lib/class/Server/command/mode.cpp:
```diff
@@ -6,8 +6,1104 @@
+ *
+ * @throw `std::exception` if a function of the C++ standard library critically fails.
+ */
+inline static int save_ban_mask_mode_to_be_cleared(
```
Why do we have a ban mask if the ban command is not implemented?

JonathanDUFOUR replied:
Because the `BAN` command doesn't exist 😆
Banning users from channels is done via the `MODE` command, and more precisely via `MODE +b nick` to add a nickname to the list of ban masks, and via `MODE -b nick` to remove a nickname from the list of ban masks.

Saphirac commented on src/lib/class/Server/command/nick.cpp:
```diff
@@ -3,100 +3,42 @@
 
-	sender.append_to_msg_out("NICK " + nickname);
-	return Success;
+	sender.append_to_msg_out(sender.prefix() + "NICK " + nickname);
```
Here we have to send the client prefix again ?

JonathanDUFOUR replied:
Yes, every time we are sending something to a client and we are doing it as a reaction to another message that comes from a client, we should add the prefix to indicate the origin of the message that we are sending. For example, the reason why we are sending a `NICK` message to the client here is because they sent us a `NICK` message first. So they are the origin of our reply. And thus, we should add their prefix.

Saphirac commented on src/lib/class/Server/command/user.cpp:
```diff
@@ -3,196 +3,79 @@
-
 /**
  * @brief
  * Finilizes the user registration process for a client,
```
Finilizes hahaha?

JonathanDUFOUR replied:
Typo, my bad, well spotted!

Saphirac commented on src/lib/class/Server/core.cpp:
```diff
@@ -3,214 +3,161 @@
-void Server::set_epoll_socket(int const epoll_socket) { this->_epoll_socket = epoll_socket; }
-void Server::set_port(int const port) { this->_port = port; }
+	tm const *const time_info = localtime(&raw_time);
+	char            buffer[42];
```
magic number ?xd

JonathanDUFOUR replied:
This one is maybe one of the only acceptable magic numbers because it has been chosen arbitrarily. It represents the size of the buffer to use to store the formatted string that will represent the current time and date. Thus, I chose 42 because it allows storing any time which has the following format: `HH:MM:SS`, and any date until `99999999999999999999999999999999999-12-31`, and of course because it's the answer to the life, the universe, and everything 🤣

Saphirac commented on src/lib/class/Server/core.cpp:
Did you remove the getters because they were unused?

JonathanDUFOUR replied:
Yes, I checked every reference to every method to remove the unused ones.

Saphirac commented on src/lib/class/Server/core.cpp:
```diff
@@ -3,214 +3,161 @@
+#include "class/exception/ProblemWithTime.hpp"
+#include <csignal>
 #include <ctime>
 #include <iostream>
```
where is iostream used?

JonathanDUFOUR replied:
Nowhere, I shall remove it, well spotted!

Saphirac commented on src/lib/class/Server/methods.cpp:
```diff
@@ -0,0 +1,186 @@
+/**
+ * @brief Starts the server, making it receiving messages, handling them and sending the appropriate replies.
+ */
+void Server::start(void)
```
I tend to think that a function calling other functions should be after theses functions in the code

JonathanDUFOUR replied:
We can do that if you want to. I used to define the methods in the same order as they appear in the header file, but I also used to define the sub-functions before the functions that call them outside of classes. So I don't mind doing the same for the methods.

Saphirac replied:
If you defined all of them in the same order as the header, we can keep it like that if everything is coherent.

Saphirac commented on src/lib/class/exception/ProblemWithClose/core.cpp:
why renaming ft_irc.hpp as ProblemWithClose() instead of creating a new file ???

JonathanDUFOUR replied:
Actually, I didn't rename it. I created the file `src/lib/class/exception/ProblemWithClose/core.cpp`, and I deleted the file `private/ft_irc.hpp` because we didn't need it anymore. And Git, instead of saving these actions as a `new file` and a `deleted file` saved them as a `renamed file` with changes inside too. It happens sometimes with Git, maybe it considers that it's lighter to save it this way in some cases.

Saphirac replied:
I see :)

Saphirac commented on src/main.cpp:
```diff
@@ -19,19 +19,18 @@ int main(int const ac, char const *const *const av)
 	if (ac != 3)
 	{
-		std::cerr << "incorrect number of args\n";
+		std::cerr << "incorrect number of arguments\n";
```
A std::endl instead of \n would be more pertinent in the case of an error.

JonathanDUFOUR replied:
As you prefer, anyway, the end of the program will flush the buffer, and this end is reached at the next line.

Saphirac replied:
Yes but it would maintain a certain logic since we always put std::endl in case of an error

Saphirac reviewed (changes requested):
Please answer my questions before merging.

JonathanDUFOUR commented on include/class/Channel.hpp:
```diff
@@ -0,0 +1,119 @@
+
+		// Methods
+		void set(ChannelMode const mode, void const *arg = NULL);
+		void clear(ChannelMode const mode, void const *arg = NULL);
```
I forgot to put the pointer as constant too.

JonathanDUFOUR commented on include/class/Channel.hpp:
```diff
@@ -0,0 +1,119 @@
+
+	// Methods
+	void set_mode(ChannelMode const mode, void const *arg = NULL);
+	void clear_mode(ChannelMode const mode, void const *arg = NULL);
```
I also forgot to put the pointer as constant here.

JonathanDUFOUR commented on include/class/Client.hpp:
```diff
@@ -3,108 +3,142 @@
-#include <string>
+#include <map>
+#include <stdint.h>
+#include <sys/epoll.h>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on include/class/Client.hpp:
```diff
@@ -3,108 +3,142 @@
 
-	std::clock_t _time_last_msg;
+		// Methods
+		void set(UserMode const mode, void const *const = NULL);
```
I forgot to name the second parameter. I shall change that.

JonathanDUFOUR commented on include/class/Server.hpp:
```diff
@@ -3,153 +3,85 @@
-#include "StatusCode.hpp"
-#include "ft_irc.hpp"
+#include "class/Client.hpp"
 #include <map>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on include/class/Server.hpp:
```diff
@@ -3,153 +3,85 @@
+#include "class/Client.hpp"
 #include <map>
 #include <netinet/in.h>
 #include <set>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on include/class/Server.hpp:
```diff
@@ -3,153 +3,85 @@
 #include <set>
-#include <string>
 #include <sys/epoll.h>
 #include <sys/socket.h>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on include/class/Server.hpp:
```diff
@@ -3,153 +3,85 @@
 #include <netinet/in.h>
 #include <set>
-#include <string>
 #include <sys/epoll.h>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on include/class/Server.hpp:
```diff
@@ -3,153 +3,85 @@
+	std::string                       _creation_date;
+	std::string                       _creation_time;
+	std::string                       _compilation_date;
+	std::string                       _compilation_time;
```
All the above fields should be constant. I will change that.

JonathanDUFOUR commented on private/class/exception/ProblemWithSettingNonBlock.hpp:
This file is unused now, I will remove it.

JonathanDUFOUR commented on private/replies.hpp:
```diff
@@ -3,18 +3,17 @@
 
 #pragma once
 
+#include "class/Client.hpp"
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on private/user_modes.hpp:
```diff
@@ -3,23 +3,28 @@
 #pragma once
 
 #include "UserMode.hpp"
 #include <cstddef>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Channel/core.cpp:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * @param name The name of the channel.
+ */
```
This doxygen comment is incorrect and should be removed. I will make this change.

JonathanDUFOUR commented on src/lib/class/Channel/methods.cpp:
```diff
@@ -0,0 +1,370 @@
+	if (include_ban_masks)
+		for (size_t i = 0; i < this->_ban_masks.size(); ++i) modes_as_string += 'b';
+
+	if (this->is_set(Limit))
```
I forgot to reuse my constant `limit_is_set`. Calling the method `Channel::is_set()` as I did instead is irrelevant. I will change that.

JonathanDUFOUR commented on src/lib/class/Client/methods.cpp:
```diff
@@ -0,0 +1,689 @@
+#include "class/exception/ProblemWithSend.hpp"
+#include "class/exception/UnknownReply.hpp"
+#include "replies.hpp"
+#include <cstdarg>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Server/command/away.cpp:
```diff
@@ -6,73 +6,37 @@
 
 #include "class/Server.hpp"
-#include "ft_irc.hpp"
+#include "class/exception/InvalidConversion.hpp"
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Server/command/away.cpp:
```diff
@@ -6,73 +6,37 @@
 #include "class/Server.hpp"
-#include "ft_irc.hpp"
+#include "class/exception/InvalidConversion.hpp"
+#include "class/exception/UnknownReply.hpp"
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Server/command/user.cpp:
```diff
@@ -3,196 +3,79 @@
 /* ************************************************************************** */
 
-#include "channel_modes.hpp"
+#include "ChannelMode.hpp"
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Server/command/user.cpp:
```diff
@@ -3,196 +3,79 @@
 #include "class/Server.hpp"
-#include "class/UserModeMask.hpp"
-#include "ft_irc.hpp"
+#include "class/specialized_string/UserModeMask.hpp"
```
This `#include` should be commented out as the class `UserModeMask` is not used anymore, as the below comment says.

JonathanDUFOUR commented on src/lib/class/Server/command/user.cpp:
```diff
@@ -3,196 +3,79 @@
+#include "class/specialized_string/UserModeMask.hpp"
 #include "replies.hpp"
-#include "user_modes.hpp"
 #include <cstdlib>
```
This #include should be commented out as the class UserModeMask is not used anymore, as the below comment says.

JonathanDUFOUR commented on src/lib/class/Server/core.cpp:
```diff
@@ -3,214 +3,161 @@
+#include "class/exception/ProblemWithStrftime.hpp"
+#include "class/exception/ProblemWithTime.hpp"
+#include <csignal>
 #include <ctime>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR commented on src/lib/class/Server/core.cpp:
```diff
@@ -3,214 +3,161 @@
+#include <csignal>
 #include <ctime>
 #include <iostream>
 #include <unistd.h>
```
Unused `#include`, I shall remove it.

JonathanDUFOUR reviewed (commented):
I forgot to remove some `#include`, and I made 1 mistake in the `Channel::Modes::to_string` method. All of this will be fixed in the next commit, before merging into the `main` branch.

JonathanDUFOUR committed:
Removed unused `#include` sections                 [OK]
Fix irrelevant call in Channel::Modes::to_string() [OK]
